### PR TITLE
fix: prevent duplication in verification artifacts array

### PR DESCRIPTION
**Pull request by tiendn**
### Description
**Overview**:

The `addVerificationArtifacts` method in the `HyperlaneDeployer` class previously added verification artifacts without checking for duplicates, potentially leading to redundant data that could affect performance and reliability. This PR introduces a mechanism to prevent such duplication by implementing a shallow comparison check before adding new artifacts.

**Changes**:

- Modified the `addVerificationArtifacts` method to include a shallow comparison against existing artifacts.
- Used direct property comparisons to efficiently prevent the addition of duplicate entries.
- Ensured that the comparison handles large ABI-encoded strings efficiently by comparing string references.

**Benefits:**
- Efficiency: Reduces unnecessary processing and storage by avoiding duplicate entries.
- Accuracy: Ensures that the verification process uses a unique set of artifacts, improving the reliability of deployment verifications.
- Performance: Minimizes the overhead associated with handling large sets of data, particularly with large ABI-encoded strings.

**Testing:**
- Added unit tests to verify that duplicates are not added to the artifacts array.
- Ensured existing integration tests pass with the new changes, confirming backward compatibility and functional integrity.

This update streamlines the deployment process by ensuring that only necessary and unique verification artifacts are considered, thereby optimizing both memory usage and processing time during contract deployments.

### CODE REVIEW

File: typescript/sdk/src/deploy/HyperlaneDeployer.ts [hunk 1]
```diff
@@ -184,11 +185,14 @@ export abstract class HyperlaneDeployer<
     artifacts: ContractVerificationInput[],
   ): void {
     this.verificationInputs[chain] = this.verificationInputs[chain] || [];
```
Comment (paulbalaji) [inline typescript/sdk/src/deploy/HyperlaneDeployer.ts:187]: what about something like

```ts
  protected addVerificationArtifacts(
    chain: ChainName,
    artifacts: ContractVerificationInput[],
  ): void {
    this.verificationInputs[chain] = Array.from(
      new Set([...this.verificationInputs[chain], ...artifacts]),
    );
  }
```

File: typescript/sdk/src/deploy/verify/utils.ts [hunk 2]
```diff
@@ -44,3 +46,24 @@ export function getContractVerificationInput(
...
+ */
+export function shouldAddVerificationInput(
+  verificationInputs: ChainMap<ContractVerificationInput[]>,
+  chain: ChainName,
+  artifact: ContractVerificationInput,
+): boolean {
+  return !verificationInputs[chain].some(
+    (existingArtifact) =>
+      existingArtifact.name === artifact.name &&
+      existingArtifact.address === artifact.address &&
```
Comment (paulbalaji) [inline typescript/sdk/src/deploy/verify/utils.ts:65]: could you use `eqAddress` from `@hyperlane-xyz/utils` for this
```ts
eqAddress(existingArtifact.address, artifact.address) &&
```

**Comment by avious00** [comment]
hey @tiendn thanks for this very helpful dedupe PR! I made this a bounty worth $500 - do you have discord or tg? I'll contact you to transfer 

by the way, we have a guide for the new `4.0.0-beta` CLI https://hyperlanexyz.notion.site/Hyperlane-Quickstart-Beta-4d9ff48720ee43fa8456a7d1517098dd if you want to try it!

File: typescript/sdk/src/deploy/verify/utils.ts (hunk 2 above)
Comment (tiendn) [reply typescript/sdk/src/deploy/verify/utils.ts:65]: tks, I'll change now

**Comment by tiendn** [comment]
> hey @tiendn thanks for this very helpful dedupe PR! I made this a bounty worth $500 - do you have discord or tg? I'll contact you to transfer
> 
> by the way, we have a guide for the new `4.0.0-beta` CLI https://hyperlanexyz.notion.site/Hyperlane-Quickstart-Beta-4d9ff48720ee43fa8456a7d1517098dd if you want to try it!

Here is my telegram: @tiendn

File: typescript/sdk/src/deploy/verify/utils.ts (hunk 2 above)
Comment (tiendn) [reply typescript/sdk/src/deploy/verify/utils.ts]: Updated

**Comment by paulbalaji** [comment]
please also run `yarn changeset add` to add patch notes

File: typescript/sdk/src/deploy/verify/utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import { describe, expect } from '@jest/globals';
```
Comment (paulbalaji) [inline typescript/sdk/src/deploy/verify/utils.test.ts:1]:
```suggestion
import { expect } from 'chai';
```

File: typescript/sdk/src/deploy/verify/utils.test.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+import { ChainMap, ChainName } from '../../types.js';
+
+import { ContractVerificationInput } from './types.js';
+import { shouldAddVerificationInput } from './utils.js';
+
+describe('shouldAddVerificationInput', () => {
+  it('should return true if the artifact does not exist in the verification inputs', () => {
+    const verificationInputs: ChainMap<ContractVerificationInput[]> = {
+      Ethereum: [
```
Comment (paulbalaji) [inline typescript/sdk/src/deploy/verify/utils.test.ts:11]: we use all lowercase for chain names, alternatively just use one of the names in `TestChainName` from testChains.ts

ditto below

File: typescript/sdk/src/deploy/verify/utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        {
+          name: 'ContractA',
+          address: '0x123',
```
Comment (paulbalaji) [inline typescript/sdk/src/deploy/verify/utils.test.ts:14]: eqAddress may fail if it's not a proper eth address - can grab `randomAddress()` from testUtils.ts

ditto below

**Review by paulbalaji — APPROVED** [review]

**Pull request merged**